README
Net::RNDC - Speak the BIND RNDC V1 Protocol

See lib/Net/RNDC.pm for example usage.

To play with Net::RNDC (Uses Dist::Zilla):

  git clone git://github.com/wolfsage/Net-RNDC.git

  cd Net-RNDC

  dzil build

WARNING: Error checking is very lax right now, so if you're key is wrong
         or BIND sends a weird response, you'll get some very interesting
         warnings. I hope to add tests and fix all of this soon :)

-- Matthew Horsfall (alh)


Changes
{{$NEXT}}

  [ENHANCEMENTS]

  * Allow consumers to pass in their own IO::Socket::INET-like object

  * Add tests for Net::RNDC functionality

0.003 Tue, Jan 08, 2013

  [BUG FIXES]

  * Replace defined-or to work with older Perls (<5.10)

0.002 Fri, Jan 04, 2013

  [BUG FIXES]

  * Fix pod spelling errors and links

  * Fix documentation for do(), reset error/response earlier in do().

  [ENHANCEMENTS]

  * Speed up Net::RNDC::Packet->parse() a bit

  * Improve the minting process (dist.ini + tests)

  * Document IPv6 support

  * More tests

0.001 Thu, Jan 3, 2013

  * First release!



t/lib/FakeSocket.pm
package
	FakeSocket; # Hide from indexer?

use strict;
use warnings;

use Carp qw(croak);

# Called by Net::RNDC::do
sub new {
	my ($self, %opt) = @_;

	$self->reset;

	for my $arg (qw(PeerAddr)) {
		unless ($opt{$arg}) {
			croak("Required arg $arg missing");
		}

		$self->{$arg} = $opt{$arg};
	}

	return $self;
}

# Actual constructor
sub create {
	my ($class, %opt) = @_;

	# Build up our $sock "package"
	my %obj;

	for my $arg (qw(on_read)) {
		unless ($opt{$arg}) {
			croak("Required arg $arg missing");
		}

		$obj{$arg} = $opt{$arg};
	}

	$obj{on_reset} = $opt{on_reset} if $opt{on_reset};

	$obj{response} = [];

	return bless \%obj, $class;
}

# Reset internal state
sub reset {
	my ($self) = @_;

	$self->{response} = [];
	$self->{PeerAddr} = '';

	$self->{on_reset} && $self->{on_reset}->($self);
}

# Fake IO::Socket::INET::send()
sub send {
	my ($self, $data) = @_;

	$self->{on_read}->($self, $data);

	return length($data);
}

# Fake IO::Socket::INET::recv()
sub recv {
	my ($self) = @_;

	if (@{$self->{response}}) {
		my $data = shift @{$self->{response}};

		$_[1] = $data;

		return '';
	}

	$@ = "No data to recv\n";
	warn "$@";

	return;
}

# Fake IO::Socket::INET::close()
sub close { }

# Add data that we'll send back to the remote end
sub append_response {
	my ($self, $data) = @_;

	push @{$self->{response}}, $data;
}

1;
__END__

=head1 NAME

FakeSocket - Pretend to be L<IO::Socket::INET>-like for testing L<Net::RNDC>.

=head1 SYNOPSIS

  my $sockpkg = FakeSocket->create(
    on_read => sub {
      my ($sock, $data) = @_;

      # Do something with data Net::RNDC just sent us
    },
  );

  my $rndc = Net::RNDC->new(sock => $sockpkg);
  $rndc->do('status');

=head1 DESCRIPTION

This package pretends to be an L<IO::Socket::INET>-like package and allows users 
to provide a hook for when the L<Net::RNDC> client sends data to this package 
thinking it's a socket.

This is useful to test L<Net::RNDC> without needing to fork and run a separate 
socket server or attempt to go async.

=head2 Constructor

=head3 create

  FakeSocket->create(%args);

Required Arguments:

=over 4

=item *

B<on_read> - A hook (subroutine) that is called with the L<FakeSocket> object 
and the data that l<Net::RNDC> called send() with. This should be the main 
driver of how L<FakeSocket> behaves in response to L<Net::RNDC>.

=back

Optional Arguments:

=over 4

=item *

B<on_reset> - A hook (subroutine) that is called each time C<< $rndc->do() >> is 
called that can be used to reset state to run new tests.

=back

=head2 Methods

=head3 append_response

  $sock->append_response($packet->data);

This should be used by tests to signal that data should be added to the outgoing 
send queue that will be read in by L<Net:RNDC>.

=head3 reset

  $sock->reset;

This is called whenever L</new> is called on an existing 
L<FakeSocket> object. It should handle resetting state for the most part, but is 
provided as part of the public interface in the event that more fine grained 
control is wanted.

=head1 SEE ALSO

t/00_rndc.t - An example of how to use this with L<Net::RNDC::Session> to mimick 
L<Net::RNDC> talking to an actual server

=head1 AUTHOR

Matthew Horsfall (alh) <WolfSage@gmail.com>

=head1 LICENSE

You may distribute this code under the same terms as Perl itself.

=cut


lib/Net/RNDC/Packet.pm
package Net::RNDC::Packet;

use strict;
use warnings;

use Net::RNDC::Exception;

use Try::Tiny;

use UNIVERSAL ();

use Carp qw(croak);
use Digest::HMAC_MD5;
use MIME::Base64 qw(decode_base64);

# lib/isccc/include/isccc/cc.h
use constant ISCCC_CCMSGTYPE_STRING     => 0x00;
use constant ISCCC_CCMSGTYPE_BINARYDATA => 0x01;
use constant ISCCC_CCMSGTYPE_TABLE      => 0x02;
use constant ISCCC_CCMSGTYPE_LIST       => 0x03;

# Serial should be created by users
my $serial = int(rand(2**32));

sub new {
	my ($class, %args) = @_;

	my @required_args = qw(
		key
	);

	my @optional_args = qw(
		version
		data
		nonce
	);

	for my $r (@required_args) {
		unless (exists $args{$r}) {
			croak("Missing required argument '$r'");
		}
	}

	if ($args{data} && (ref($args{data}) || '' ) ne 'HASH') {
		croak("Argument 'data' must be a HASH");
	}

	if (exists $args{version} && ($args{version} || '') !~ /^\d+\z/) {
		croak("Argument 'version' must be a number");
	}

	if (exists $args{nonce} && ($args{nonce} || '') !~ /^\d+\z/) {
		croak("Argument 'nonce' must be a number");
	}

	my %object = (
		key => $args{key},
		data => {
			_ctrl => {
				_ser => $serial++,
			},
		},
		version => $args{version} || 1,
	);

	if ($args{data}) {
		$object{data}{_data} = $args{data};
	} else {
		$object{data}{_data}{type} = undef;
	}

	if ($args{nonce}) {
		$object{data}{_ctrl}{_nonce} = $args{nonce};
	}

	return bless \%object, $class;
}

sub parse {
	my ($self, $data) = @_;

	$self->_set_error('');

	unless ($self->_cklen($data, 55)) {
		return 0;
	}

	# Everything after first 51 bytes is what needs to be signed
	my $buff = substr($data, 51);

	my $length = unpack('N', $data);
	$data = substr($data, 4);

	unless ($self->_cklen($data, $length)) {
		return 0;
	}

	my $version = unpack('N', $data);
	$data = substr($data, 4);

	unless ($version == 1) {
		return $self->_set_error("Unknown protocol version '$version'");
	}

	my ($aauth, $check);

	try {
		$data = _table_fromwire(\$data);

		$aauth = $data->{_auth}{hmd5};

		$check = $self->_sign($buff);
	} catch {
		my $err = $_;

		if (UNIVERSAL::isa($err, 'Net::RNDC::Exception')) {
			$self->_set_error($err);
		} else {
			die $err;
		}
	};

	return 0 if $self->error;

	if ($check ne $aauth) {
		return $self->_set_error("Couldn't validate packet with provided key");
	}

	try {
		$self->{data} = _table_fromwire(\$buff);
	} catch {
		my $err = $_;

		if (UNIVERSAL::isa($err, 'Net::RNDC::Exception')) {
			$self->_set_error($err);
		} else {
			die $err;
		}
	};

	return 0 if $self->error;

	$self->_set_error($self->{data}->{_data}{err});

	return $self->error ? 0 : 1;
}

# Set an error. Uses Net::RNDC::Exception to get file/line number
sub _set_error {
	my ($self, $error) = @_;

	if (!$error) {
		$self->{error} = '';
	} elsif (UNIVERSAL::isa($error, 'Net::RNDC::Exception')) {
		$self->{error} = $error->error;
	} else {
		my $e = Net::RNDC::Exception->new($error);
		$self->{error} = $e->error;
	}

	return 0;
}

# Return error string if any
sub error {
	my ($self) = @_;

	return $self->{error};
}

# Return packet data in binary form
sub data {
	my ($self) = @_;

	$self->_set_error('');

	$self->{data}->{_ctrl}->{_tim} = time;
	$self->{data}->{_ctrl}->{_exp} = time + 60;

	my ($udata, $cksum, $wire);

	try {
		$udata = $self->_unsigned_data;

		$cksum = $self->_sign($udata);

		$wire = _table_towire({
			_auth => {
				hmd5 => $cksum,
			},
		}, 'no_header');
	} catch {
		my $err = $_;

		if (UNIVERSAL::isa($err, 'Net::RNDC::Exception')) {
			$self->_set_error($err);
		} else {
			die $err;
		}
	};

	return if $self->error;

	$wire .= $udata;

	return pack('N', length($wire) + 4) . pack('N', $self->{version}) . $wire;
}

# Return the table of data to be signed
sub _unsigned_data {
	my ($self) = @_;

	return _table_towire($self->{data}, 'no_header');
}

# Sign data with our key, return digest
sub _sign {
	my ($self, $data) = @_;

	my $hmac = Digest::HMAC_MD5->new(decode_base64($self->{key}));

	$hmac->add($data);

	return $hmac->b64digest;
}

# Take a string from binary format and return it
sub _binary_fromwire {
	my ($wire) = @_;

	my $data = $$wire;
	$$wire = '';

	return $data;
}

# Pack a string into its binary representation
sub _binary_towire {
	my ($data) = @_;

	if (!defined $data) {
		$data = 'null';
	}

	return pack('c', ISCCC_CCMSGTYPE_BINARYDATA)
	     . pack('N', length($data))
	     . ($data);
}

# Take a table from binary format and return a hashref
sub _table_fromwire {
	my ($wire) = @_;

	my %table;

	while ($$wire) {
		_cklen_d($$wire, 1);
		my $key_len = unpack('c', $$wire);
		$$wire = substr($$wire, 1);

		_cklen_d($$wire, $key_len);
		my $key = substr($$wire, 0, $key_len);
		$$wire = substr($$wire, $key_len);

		$table{$key} = _value_fromwire($wire);
	}

	return \%table;
}

# Pack a hashref into its binary representation
sub _table_towire {
	my ($data, $no_header) = @_;

	my $table;

	for my $k (sort keys %$data) {
		$table .= pack('c', length($k));
		$table .= $k;
		$table .= _value_towire($data->{$k});
	}

	if ($no_header) {
		return $table;
	} else {
		my $msg_type = pack('c', ISCCC_CCMSGTYPE_TABLE);
		return $msg_type . pack('N', length($table)) . $table;
	}
}

# Take a list from binary representation and return an arrayref
sub _list_fromwire {
	my ($wire) = @_;

	my @list;
	while ($$wire) {
		push @list, _value_fromwire($wire);
	}

	return \@list;
}

# Pack an arrayref into its binary representation
sub _list_towire {
	my ($data) = @_;

	my $msg_type = pack('c', ISCCC_CCMSGTYPE_LIST);
	my $list;

	for my $d (@$data) {
		$list .= _value_towire($d);
	}

	return $msg_type . pack('N', length($list)) . $list;
}

# Take a value, whatever it may be, and unpack it into perl data types
sub _value_fromwire {
	my ($wire) = @_;

	_cklen_d($$wire, 5);

	my $msg_type = unpack('c', $$wire);
	$$wire = substr($$wire, 1);

	my $len = unpack('N', $$wire);
	$$wire = substr($$wire, 4);

	_cklen_d($$wire, $len);
	my $data = substr($$wire, 0, $len);
	$$wire = substr($$wire, $len);

	if ($msg_type == ISCCC_CCMSGTYPE_BINARYDATA) {
		return _binary_fromwire(\$data);
	} elsif ($msg_type == ISCCC_CCMSGTYPE_TABLE) {
		return _table_fromwire(\$data);
	} elsif ($msg_type == ISCCC_CCMSGTYPE_LIST) {
		return _list_fromwire(\$data);
	} else {
		die Net::RNDC::Exception->new(
			"Unknown message type '$msg_type' in _value_fromwire"
		);
	}
}

# Take a perl data structure and pack it into binary format
sub _value_towire {
	my ($data) = @_;

	my $r = ref $data || 'binary';

	if ($r eq 'HASH') {
		return _table_towire($data);
	} elsif ($r eq 'ARRAY') {
		return _list_towire($data);
	} elsif ($r eq 'binary') {
		return _binary_towire($data);
	} else {
		die Net::RNDC::Exception->new(
			"Unknown data type '$r' in _value_towire"
		);
	}
}

# Sets an error and returns 0 if the buff isn't at least $len bytes
# unless ($self->_cklen($buff, $len)) {
#	return 0;
# }
sub _cklen {
#	my ($self, $buff, $len) = @_;

	unless ((length($_[1]) || 0) >= $_[2]) {
		$_[0]->_set_error(Net::RNDC::Exception->new(
			"Unexpected end of data reading buffer. (Expected $_[2] more bytes at least)"
		));

		return 0;
	}

	return 1;
}

# Throws an exception if the buff isn't at least $len bytes
#
# _cklen_d($buff, $len)
sub _cklen_d {
#	my ($buff, $len) = @_;

	unless ((length($_[0]) || 0) >= $_[1]) {
		die Net::RNDC::Exception->new(
			"Unexpected end of data reading buffer. (Expected $_[1] more bytes at least)"
		);
	}
}

1;
__END__

=head1 NAME

Net::RNDC::Packet - RNDC Protocol V1 Packet Parsing and Generation

=head1 SYNOPSIS

To send an RNDC command and get a response:

  use IO::Socket::INET;
  use Net::RNDC::Packet;

  my $buff;
  my $key = 'aabc';

  my $c = IO::Socket::INET->new(
    PeerAddr => '127.0.0.1:953',
  ) or die "Failed to create a socket: $@ ($!)";

  # Send opener packet
  my $pkt = Net::RNDC::Packet->new(
    key => $key,
  );

  $c->send($pkt->data);

  # Read nonce response
  $c->recv($buff, 4096);

  $pkt->parse($buff);

  # Send command request with nonce
  my $nonce = $pkt->{data}->{_ctrl}{_nonce};

  my $cmd = Net::RNDC::Packet->new(
    key => $key,
    nonce => $nonce,
    data => {type => 'status'},
  );

  $c->send($cmd->data);

  # Read final response
  $c->recv($buff, 4096);

  $cmd->parse($buff);

  my $resp = $cmd->{data}{_data}{text} || 'command success';

  print "$resp\n";

=head1 DESCRIPTION

This package provides low-level RNDC V1 protocol parsing and generation. It 
allows full control over the data in the sent/received packets.

Currently this is provided by hacking at C<< $pkt->{data} >>, setter/getter 
methods will be forthcoming.

=head2 Constructor

=head3 new

  my $packet = Net::RNDC::Packet->new(%args);

Arguments:

=over 4

=item *

B<key> - The Base64 encoded HMAC-MD5 key to sign/verify packets with.

=item *

B<data> - A hashref of data to put in the request of the packet. Currently, BIND 
only understand commands in the C<type> key. For example:

  data => { type => 'status' },

=item *

B<nonce> - The nonce data returned from the remote nameserver. Located in the 
parsed packet in the _ctrl section:

  nonce => $packet->{data}->{_ctrl}{_nonce},

=back

=head2 Methods

=head3 data

  my $binary = $packet->data;

Generates a binary representation of the packet, suitable for sending over the 
wire.

=head3 parse

  if ($packet->parse($binary)) { ... }

Parses data from the wire and populates the current packet with the information, 
as well as verifies the data with the provided B<key> that was passed to the 
constructor. Returns 1 on success, 0 on failure. Check L</error> if there's a 
failure.

=head3 error

  my $err = $packet->error;

Returns a string error, if any, after packet parsing or generation failed.

=head1 TODO

=over 4

=item *

Methods for modifying the different data parts of an RNDC message

=back

=head1 SEE ALSO

L<Net::RNDC> - Simple RNDC communication.

L<Net::RNDC::Session> - Manage the 4-packet RNDC session

=head1 AUTHOR

Matthew Horsfall (alh) <WolfSage@gmail.com>

=head1 LICENSE

You may distribute this code under the same terms as Perl itself.

=cut


lib/Net/RNDC/Session.pm
package Net::RNDC::Session;

use strict;
use warnings;

use Net::RNDC::Packet;

use Carp qw(croak);

# Controls the flow in next(). undef means next() should never
# be called if we've reached this state
my %states = (
	start        => '_got_start',
	want_read    => '_got_read',
	want_write   => '_got_write',
	want_error   => undef,
	want_finish  => undef,
);

sub new {
	my ($class, %args) = @_;

	my @required_subs = qw(
		want_read
		want_write
		want_finish
		want_error
	);

	my @optional_subs = qw(
	);

	my @required_args = qw(
		key
		command
	);

	my @optional_args = qw(
		is_client
		is_server
	);

	for my $r (@required_subs, @required_args) {
		unless (exists $args{$r}) {
			croak("Missing required argument '$r'");
		}
	}

	unless (exists $args{is_client} || exists $args{is_server}) {
		croak("Argument 'is_client' or 'is_server' must be defined");
	}

	for my $r (@required_subs, @optional_subs) {
		next unless exists $args{$r};

		unless ((ref $args{$r} || '') eq 'CODE') {
			croak("Argument '$r' is not a code ref");
		}
	}

	if (exists $args{is_client} && exists $args{is_server}) {
		croak("Argument 'is_client' cannot be mixed with 'is_server'");
	}

	my %obj = map {
		$_ => $args{$_}
	} grep { exists $args{$_} } (@required_subs, @optional_subs, @required_args, @optional_args);

	if (exists $args{is_client}) {
		$obj{is_client} = 1;
	} else {
		$obj{is_server} = 1;
	}

	my $obj = bless \%obj, $class;

	# Base state
	$obj->_init;

	return $obj;
}

# Maybe open up to public as reset()?
sub _init {
	my ($self) = @_;

	# Have we sent our syn/ack opener?
	$self->{nonce} = 0;

	$self->_state('start');
}

# Set/retrieve state
sub _state {
	my ($self, $state) = @_;

	if ($state) {
		unless (exists $states{$state}) {
			croak("Unknown state $state requested");
		}

		$self->{state} = $state;
	}

	return $self->{state};
}

sub _is_client { return $_[0]->{'is_client'} }
sub _is_server { return $_[0]->{'is_server'} }
sub _key       { return $_[0]->{'key'}       }
sub _nonce     { return $_[0]->{'nonce'}     }
sub _command   { return $_[0]->{'command'}   }

# Entry point. Always.
sub start {
	my ($self) = @_;

	unless (my $state = $self->_state eq 'start') {
		croak("Attempt to re-use an existing session in state '$state'");
	}

	$self->next;
}

# Move things along. Pass in data if needed
sub next {
	my ($self, $data) = @_;

	my $sub = $states{$self->_state};

	unless ($sub) {
		croak("next() called on bad state '" . $self->_state . "'");
	}

	$self->$sub($data);

	return;
}

# _got subs are called after a want_* sub has been called and next() has been used

# Starting out
sub _got_start {
	my ($self, $data) = @_;

	if ($self->_is_client) {
		# Client step 1: send a request packet with no data section
		my $packet = Net::RNDC::Packet->new(
			key => $self->_key,
		);

		$self->_state('want_write');

		return $self->_run_want('want_write', $packet->data, $packet);
	} else {
		# Server step 1: expect a packet with no data section
		$self->_state('want_read');

		return $self->_run_want('want_read');
	}
}

sub _got_read {
	my ($self, $data) = @_;

	if ($self->_is_client) {
		my $packet = Net::RNDC::Packet->new(key => $self->_key);

		if (!$packet->parse($data)) {
			$self->_state('want_error');

			return $self->_run_want('want_error', $packet->error);
		}

		if (! $self->_nonce) {
			# Client step 2: Parse response, get nonce
			$self->{nonce} = 1;

			my $nonce = $packet->{data}->{_ctrl}{_nonce};

			# Client step 3: Send request with nonce/data section
			my $packet2 = Net::RNDC::Packet->new(
				key => $self->_key,
				nonce => $nonce,
				data => {type => $self->_command},
			);

			$self->_state('want_write');

			return $self->_run_want('want_write', $packet2->data, $packet2);
		} else {
			# Client step 4: Read response to command
			my $response = $packet->{data}{_data}{text} || 'command success';

			$self->_state('want_finish');

			return $self->_run_want('want_finish', $response);
		}
	} else {
		my $packet = Net::RNDC::Packet->new(key => $self->_key);

		if (!$packet->parse($data)) {
			$self->_state('want_error');

			return $self->_run_want('want_error', $packet->error);
		}

		if (! $self->_nonce) {
			$self->{nonce} = 1;

			my $nonce = int(rand(2**32));

			$self->{_nonce_data} = $nonce;

			my $challenge = Net::RNDC::Packet->new(
				key => $self->_key,
				nonce => $nonce,
			);

			$self->_state('want_write');

			return $self->_run_want('want_write', $challenge->data, $challenge);
		} else {
			my $nonce = $self->{_nonce_data};

			# TODO: Add time/expiry checking
			# Invalid: (_tim + clockskew < now || _tim - clockskew > now)
			# Invalid: now > exp
			# Also check serial?

			unless ($packet->{data}->{_ctrl}{_nonce}) {
				$self->_state('want_error');

				return $self->_run_want('want_error', "Client nonce not set");	
			}

			unless ($packet->{data}->{_ctrl}{_nonce} == $nonce) {
				$self->_state('want_error');

				return $self->_run_want('want_error', "Client nonce does not match");
			}

			my $response = Net::RNDC::Packet->new(
				key => $self->_key,
				data => {text => $self->_command},
			);

			$self->_state('want_write');

			$self->_run_want('want_write', $response->data, $response);

			$self->_state('want_finish');

			$self->_run_want('want_finish');
		}
	}
}

sub _got_write {
	my ($self) = @_;

	# As a client, after every write we expect a read
	if ($self->_is_client) {
		$self->_state('want_read');

		return $self->_run_want('want_read');
	} elsif ($self->_is_server) {
		$self->_state('want_read');

		return $self->_run_want('want_read');
	}
}

# Run the requested want_* sub
sub _run_want {
	my ($self, $sub, @args) = @_;

	my $ref = $self->{$sub};

	$ref->($self, @args);
}

1;
__END__

=head1 NAME

Net::RNDC::Session - Helper package to manage the RNDC 4-packet session

=head1 SYNOPSIS

To use synchronously as a client:

  use IO::Socket::INET;
  use Net::RNDC::Session;

  my $c = IO::Socket::INET->new(
    PeerAddr => '127.0.0.1:953',
  ) or die "Failed to create a socket: $@ ($!)";

  # Our response
  my $response;

  my $session = Net::RNDC::Session->new(
    key         => 'abcd',
    command     => 'status',
    is_client   => 1,

    want_write =>  sub { my $s = shift; $c->send(shift); $s->next; },
    want_read  =>  sub { my $s = shift; my $b; $c->recv($b, 4096); $s->next($b); },
    want_finish => sub { my $s = shift; $response = shift; },
    want_error =>  sub { my $s = shift; my $err = shift; die "Error: $err\n"; },
  );

  # Since we call next() in want_read/want_write above, this will do everything
  $session->start;

  print "Response: $response\n";

To use asynchronously (for example, with IO::Async):

TBD

To use as a server:

TBD

To use asynchronously as a server:

TBD

=head1 DESCRIPTION

This package is intended to provide the logic for an RNDC client session which 
can used  to run a single command against a remote server and get a response.
See L</SESSION> below for a description of the RNDC client session logic.

This package also supports running sessions as an RNDC server.

For simple use of the RNDC protocol, see L<Net::RNDC>.

There is no socket logic here, that must be provided to this class through the 
constructor in the various C<want_*> methods. This allows for 
synchronous/asynchronous use with a little work.

This package does generate and parse L<Net::RNDC::Packet>s, but the 
L</want_read> and L</want_write> methods allow you to peak at this data before 
it's parsed and before it's sent to the remote end to allow slightly more 
fine-grained control.

To manage the entire process yourself, use L<Net::RNDC::Packet>.

=head1 SESSION

An RNDC client session (where one is sending commands to a remote nameserver 
expecting a response) contains 4 packets.

All packets contain a timestamp/expiracy timestamp to denote a packet validity 
window, as well as an HMAC-MD5 signature of the packets data using a shared 
private key, and a serial number to identify the packet.

=over 4

=item 1

  CLIENT->send(<opening packet>)

The opening packet contains a '_data' section with an undef 'type'.

=item 2

  SERVER->send(<nonce packet>)

The server response packet contains a 'nonce' integer which should be 
copied into the next request.

=item 3

  CLIENT->send(<command packet>)

The nonce should be included in the command packet in the '_ctrl' section, and 
the command to be run on the remote section should be in the 'type' parameter of 
the '_data' section.

=item 4

  SERVER->send(<response packet>)

The response packet will contain an 'error' parameter in the '_data' section if 
something went wrong, otherwise the response will be in the 'text' parameter of 
the '_data' section.

=back

If at any time the remote end disconnects prematurely, this may indicate any of 
the following (along with normal network issues):

=over 4

=item *

The clocks are off

=item *

The key is incorrect

=item *

The window has expired

=back

=head1 SEE ALSO

L<Net::RNDC> - Simple RNDC communication.

L<Net::RNDC::Packet> - Low level RNDC packet manipulation.

=head1 AUTHOR

Matthew Horsfall (alh) <WolfSage@gmail.com>

=head1 LICENSE

You may distribute this code under the same terms as Perl itself.

=cut


lib/Net/RNDC/Exception.pm
package Net::RNDC::Exception;

use strict;
use warnings;

sub new {
	my ($class, $error, $level) = @_;

	$level ||= 1;

	my (undef, $file, $line) = caller($level);

	return bless {
		error => $error,
		file  => $file,
		line  => $line,
	}, $class;
}

sub error {
	my ($self) = @_;

	return "'$self->{error}' at '$self->{file}' line '$self->{line}'";
}

1;
__END__

=head1 NAME

Net::RNDC::Exception - Internal exception class

=head1 SYNOPSIS

For use within Net::RNDC* only.

To throw an error:

  die Net::RNDC::Exception->new("some error");

Typically used with L<Try::Tiny>:

  try {
    something_that_dies_like_above();
  } catch {
    my $err = $_;
    if (UNIVERSAL::isa($err, 'Net::RNDC::Exception')) {
      warn "Error: " . $err->error . "\n";
    }
  };

=head1 DESCRIPTION

This package is used to pass exceptions around internally so that a simple error 
interface can be exposed by the API and any actual exceptions are real problems 
with usage/bugs.

=head2 Constructor

=head3 new

  Net::RNDC::Exception->new($error);

Constructs a new exception with the given B<$error> and attached file/line 
number information to it.

Required Arguments:

=over 4

=item *

B<$error> - A string describing the error.

=back

=head2 Methods

=head3 Error

  my $error = $exception->error;

Returns a string containing the error from object construction, along with a 
file and line identifier.

=head1 AUTHOR

Matthew Horsfall (alh) <WolfSage@gmail.com>

=head1 LICENSE

You may distribute this code under the same terms as Perl itself.

=cut


lib/Net/RNDC.pm
package Net::RNDC;
# ABSTRACT: Speak the BIND RNDC protocol

use strict;
use warnings;

use UNIVERSAL ();

use Carp qw(croak);

use Net::RNDC::Session;

my $Sock;

BEGIN {
	eval 'use IO::Socket::INET6;';

	if ($@) {
		eval 'use IO::Socket::INET;';

		die $@ if $@;

		$Sock = 'IO::Socket::INET';
	} else {
		$Sock = 'IO::Socket::INET6';
	}
}

# Required for new()
my @required_args = qw(
);

# Optional for new()/do()
my @optional_args = qw(
	key
	host
	port
	sock
);

sub new {
	my ($class, %args) = @_;

	my %obj = $class->_parse_args(%args);

	return bless \%obj, $class;
}

sub _parse_args {
	my ($class, %args) = @_;

	for my $r (@required_args) {
		unless ($args{$r}) {
			croak("Missing required argument '$r'");
		}
	}

	$args{port} ||= 953;
	$args{sock} ||= $Sock;

	return map {
		$_ => $args{$_}
	} grep { $args{$_} } (@required_args, @optional_args);
}

sub _check_do_args {
	my ($self, %args) = @_;

	for my $r (qw(key host)) {
		unless ($args{$r}) {
			croak("Missing required argument '$r'");
		}
	}

	unless(UNIVERSAL::can($args{sock}, 'new')) {
		croak("Package '$args{sock}' has no 'new' method");
	}	
}

sub do {
	my ($self, $command, %override) = @_;

	$self->{response} = $self->{error} = '';

	my $host = $self->{host};
	my $port = $self->{port};
	my $key  = $self->{key};
	my $sock = $self->{sock};

	if (%override) {
		my %args = $self->_parse_args(
			host => $host,
			port => $port,
			key  => $key,
			sock => $sock,
			%override,
		);

		$host = $args{host};
		$port = $args{port};
		$key  = $args{key};
		$sock = $args{sock};
	}

	$self->_check_do_args(
		host => $host,
		port => $port,
		key  => $key,
		sock => $sock,
	);

	my $c = $sock->new(
		PeerAddr => "$host:$port",
	);

	unless ($c) {
		$self->{error} = "Failed to create a $sock: $@ ($!)";

		return 0;
	}

	for my $meth (qw(send recv close)) {
		unless (UNIVERSAL::can($c, $meth)) {
			croak("Object returned from '$sock->new()' has no '$meth' method");
		}
	}

	# Net::RNDC::Session does all of the work
	my $sess = Net::RNDC::Session->new(
		key         => $key,
		command     => $command,
		is_client   => 1,

		want_write => sub {
			my $s = shift;

			$c->send(shift);

			$s->next;
		},

		want_read => sub {
			my $s = shift;

			my $buff;

			$c->recv($buff, 4096);

			$s->next($buff);
		},

		want_finish => sub {
			my $s = shift;
			my $res = shift;

			$self->{response} = $res;
		},

		want_error => sub {
			my $s = shift;
			my $err = shift;

			$self->{error} = $err;
		}
	);

	# Work!
	$sess->start;

	$c->close;

	if ($self->response) {
		return 1;
	} else {
		return 0;
	}
}

sub response {
	my ($self) = @_;

	return $self->{response};
}

sub error {
	my ($self) = @_;

	return $self->{error};
}

1;
__END__;

=head1 NAME

Net::RNDC - Speak the BIND Remote Name Daemon Control (RNDC) V1 protocol

=head1 SYNOPSIS

Simple synchronous command/response:

  use Net::RNDC;

  my $rndc = Net::RNDC->new(
    host => '127.0.0.1',
    port => 953,         # Defaults to 953
    key  => 'abcd',
  );

  if (!$rndc->do('status')) {
    die "RNDC failed: " . $rndc->error;
  }

  print $rndc->response;

All arguments to new() are allowed in do:

  my $rndc = Net::RNDC->new();

  my $key = 'abcd';

  for my $s (qw(127.0.0.1 127.0.0.2)) {
    if (!$rndc->do('status', key => $key, host => $s)) {
      my $err = $rndc->error;
    } else {
      my $resp = $rndc->response;
    }
  }

=head1 DESCRIPTION

This package provides a synchronous, easy to use interface to the RNDC V1 
protocol. For more mid-level control, see L<Net::RNDC::Session>, and for 
absolute control, L<Net::RNDC::Packet>.

=head2 Constructor

=head3 new

  Net::RNDC->new(%args);

Optional Arguments:

=over 4

=item *

B<key> - The Base64 encoded HMAC-MD5 private key to use.

=item *

B<host> - The hostname/IP of the remote server to connect to. If 
L<IO::Socket::INET6> is installed, IPv6 support will be enabled.

=item *

B<port> - The port to connect to. Defaults to I<953>.

=item *

B<sock> - A package, like L<IO::Socket::INET>, that provides a C<new> function 
which returns an object capable of performing C<send>, C<recv>, and C<close>, 
and behaving similar to them. See C<< perldoc -f >> for each of the methods 
above to see how they should behave. Defaults to L<IO::Socket::INET6> if 
available, otherwise C<IO::Socket::INET>. The C<new> function should accept one 
parameter - B<PeerAddr> - which will be a I<hostname:port> string.

=back

=head2 Methods

=head3 do

  $rndc->do($command);

  $rndc->do($commands, %args);

Connects to the remote nameserver configured in L</new> or passed in to  
B<%args> and sends the specified command.

Returns 1 on success, 0 on failure.

Arguments:

=over 4

=item *

B<$command> - The RNDC command to run. For example: C<status>.

=back

Optional Arguments - See L</new> above.

=head3 error

  $rndc->error;

Returns the last string error from a call to L</do>, if any. Only set if 
L</do> returns 0.

=head3 response

  $rndc->response;

Returns the last string response from a call to L</do>, if any. Only set if 
L</do> returns 1.

=head1 SEE ALSO

L<Net::RNDC::Session> - Manage the 4-packet RNDC session

L<Net::RNDC::Packet> - Low level RNDC packet manipulation.

=head1 AUTHOR

Matthew Horsfall (alh) <WolfSage@gmail.com>

=head1 LICENSE

You may distribute this code under the same terms as Perl itself.

=cut
